client/tests/tests/events.rs

```diff
@@ -6,49 +6,100 @@ use std::{
-                if let Ok(Event::Pipeline(event)) = event {
-                    //TODO: check transaction hash
-                    if event.entity_type == PipelineEntityType::Transaction
-                        && event.status == PipelineStatus::Committed
-                    {
-                        *committed_event_received_clone
-                            .write()
-                            .expect("Failed to acquire lock.") = true;
-                    }
-                }
-            },
-        );
-    });
-    thread::sleep(pipeline_time * 2);
-    //When
-    iroha_client.submit(create_asset)?;
-    thread::sleep(pipeline_time * 2);
-    //Then
-    assert!(*committed_event_received.read().unwrap());
+    test_with_instruction_and_status(create_asset, true)?;
+    let fail = FailBox::new("Failing transaction to test Rejected event.");
+    test_with_instruction_and_status(fail, false)
+}
+
+#[allow(clippy::needless_range_loop)]
+fn test_with_instruction_and_status(
+    instruction: impl Into<Instruction>,
+    should_be_committed: bool,
+) -> Result<()> {
+    let instruction: Instruction = instruction.into();
+    for submitting_peer in 0..PEER_COUNT {
+        let (_rt, network, _) =
+            <Network>::start_test_with_runtime(PEER_COUNT.try_into().unwrap(), 1);
+        let pipeline_time = Configuration::pipeline_time();
+
+        // Given
+        thread::sleep(pipeline_time * 4);
+
+        let committed_event_received = Arc::new(RwLock::new([false; PEER_COUNT]));
+        let validating_event_received = Arc::new(RwLock::new([false; PEER_COUNT]));
+        let rejected_event_received = Arc::new(RwLock::new([false; PEER_COUNT]));
+        let peers: Vec<_> = network.peers().collect();
+        let mut submitter_client = Client::test(&peers[submitting_peer].api_address);
+        let transaction = submitter_client
+            .build_transaction(vec![instruction.clone()], UnlimitedMetadata::new())?;
+        for receiving_peer in 0..PEER_COUNT {
+            let committed_event_received_clone = committed_event_received.clone();
+            let validating_event_received_clone = validating_event_received.clone();
+            let rejected_event_received_clone = rejected_event_received.clone();
+            let listener_client = Client::test(&peers[receiving_peer].api_address);
+            let hash = transaction.hash();
+            let _handle = thread::spawn(move || {
+                listener_client.for_each_event(
+                    EventFilter::Pipeline(PipelineEventFilter::by_entity(
+                        PipelineEntityType::Transaction,
+                    )),
+                    |event| {
+                        if let Ok(Event::Pipeline(event)) = event {
+                            if event.entity_type == PipelineEntityType::Transaction
+                                && event.hash == *hash
+                            {
+                                if event.status == PipelineStatus::Committed {
```

Comment:
These branches should be mutually exclusive, so I would use `if-else`. Even better I would use match with arms instead of `if-else` because it wouldn't compile if another variant is added to the enum.